### PR TITLE
#200 - Update Item Properties from Description

### PR DESCRIPTION
**Pull request by kovax**

### CODE REVIEW

File: kernel/src/main/java/org/cristalise/kernel/property/PropertyArrayList.java [hunk 1]
```diff
@@ -44,10 +44,27 @@ public PropertyArrayList(ArrayList<Property> aList) {
 
     public void put(Property p) {
         //TODO: this shall be based on the contains() of the list
+        if (list.contains(p)) {
+
```
Comment (taurus227) [inline kernel/src/main/java/org/cristalise/kernel/property/PropertyArrayList.java]: no contents?

File: kernel/src/main/java/org/cristalise/kernel/property/PropertyUtility.java [hunk 2]
```diff
@@ -157,4 +172,17 @@ static public CastorHashMap convertTransitiveProperties(PropertyDescriptionList
         }
         return props;
     }
+
+    public static void writeProperty(ItemPath item, String name, String value, Object locker)
+            throws PersistencyException, ObjectCannotBeUpdated, ObjectNotFoundException
+    {
+        Property prop = (Property) Gateway.getStorage().get(item, ClusterType.PROPERTY + "/" + name, locker);
+
```
Comment (taurus227) [inline kernel/src/main/java/org/cristalise/kernel/property/PropertyUtility.java]: Shouldn't we first insert here `if value.equals(prop.getValue()) return;` ?

File: kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                    Property initProp = initProps.get(existingPropName);
+
+                    if (initProp != null) {
+                        existingProp.setValue(initProp.getValue());
+                        Gateway.getStorage().put(item, existingProp, locker);
+                    }
+                }
+                else {
+                    //update existing immutable Property if its default value was changed
+                    String defaultValue = PropertyUtility.getValue(propDesc.list, existingPropName);
```
Comment (taurus227) [inline kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java]: why does this return the "default" value?

File: kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+                    if (StringUtils.isNotBlank(defaultValue) && !defaultValue.equals(existingProp.getValue())) {
+                        existingProp.setValue(defaultValue);
```
Comment (taurus227) [inline kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java]: so if the existingValue is not the default value, we set it to default? why?

File: kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java (hunk 3 above)
Comment (kovax) [reply kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java]: this method is part of the API, so we needed it for backward compability

File: kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java (hunk 4 above)
Comment (taurus227) [reply kernel/src/main/java/org/cristalise/kernel/lifecycle/instance/predefined/UpdateProperitesFromDescription.java]: If the default value was changed in the definition, the immutable property should have that value.

**Pull request merged**